test/Generator/ParameterGeneratorTest.php

```diff
@@ -576,6 +577,23 @@ public function testSetGetVariadic()
         self::assertSame('$foo', $parameter->generate());
     }
 
+    public function testGeneratingVariadicParameterWithDefaultValueThrowsInvalidArgumentException(): void
+    {
+        $parameter = new ParameterGenerator();
+
+        $parameter->setName('parameter');
+        $parameter->setType('int');
+        $parameter->setPosition(1);
+        $parameter->setVariadic(true);
+
+        $parameter->setDefaultValue([]);
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Variadic parameter cannot have a default value');
```

Comment:
I wonder if the exception should be raised by `->generate()` or by `->setDefaultValue()` and `->setVariadic()`? :thinking:

Reply:
@Ocramius 
I also considered it this alternative.

The reason I went with `generate` is that: `generate()` is the central place that takes all the properties that were previously set and combines them to generate the output. It is always called and it is called last. 
So it is a good central place to validated if those properties make sense together.

If we put the validate in `->setDefaultValue()` and `->setVariadic()`, the two setters have to validate each other.
We would have to duplicate the check, because we do not know the order in which the two method would be called.
The validation logic is also would be distributed across the setters, instead of being in one central place.
And the two setters are now coupled to each other.
Thus we get more complexity.

Reply:
> The validation logic is also would be distributed across the setters, instead of being in one central place.
 > And the two setters are now coupled to each other.
 > Thus we get more complexity.

Agree that it leads to more complexity, but it is still better than having a crash much later in the process.

To envision this more clearly:

```php
$parameterGenerator = someThirdPartyComponentCall();

// 10000 rows of code here

$parameterGenerator->generate(); // crash here - you will never understand what caused it
```

Therefore, I'd say that we need to rework this to include the exception in both setters: ugly, but necessary.

Reply:
@Ocramius 

Makes sense.
Applied the changes.

Please check again.